File: cpgateway/app/api/endpoints/system_settings.py
"""
系统设置管理接口（仅 SuperAdmin）。
GET /api/v1/system/settings       — 获取全部设置（secret 字段脱敏）
PUT /api/v1/system/settings       — 批量更新（跳过值为 "******" 的 secret 字段）
POST /api/v1/system/settings/test-notification — 测试指定渠道连通性
"""
import json
import smtplib
import email.mime.text
import ssl
import asyncio
import hmac as hmac_lib
import hashlib
import base64
import time

import httpx
from fastapi import APIRouter, Body, Depends, HTTPException, status, BackgroundTasks
from sqlalchemy.ext.asyncio import AsyncSession
from typing import Any, Dict, List

from app.core.database import get_db, AsyncSessionLocal
from app.api.deps import get_current_superuser
from app.models.user import User
from app.schemas.system_setting import (
    SettingResponse,
    SettingsListResponse,
    TestNotificationRequest,
    TestNotificationResponse,
    MASKED,
)
from app.services.settings_service import settings_service, SETTINGS_METADATA, _deserialize
from app.core.logging_config import logger

router = APIRouter()


@router.get("", response_model=SettingsListResponse)
async def list_settings(
    db: AsyncSession = Depends(get_db),
    _: User = Depends(get_current_superuser),
):
    """获取所有系统设置（含元数据默认值补全，secret 字段脱敏）"""
    rows = await settings_service.get_all(db)
    return SettingsListResponse(
        settings=[SettingResponse.model_validate(r) for r in rows]
    )


@router.put("", response_model=SettingsListResponse)
async def update_settings(
    background_tasks: BackgroundTasks,
    payload: Dict[str, Any] = Body(..., description="{ setting_key: new_value, ... }"),
    db: AsyncSession = Depends(get_db),
    _: User = Depends(get_current_superuser),
):
    """
    批量更新系统设置。
    payload: { "SMTP_HOST": "mail.example.com", "SMTP_PORT": 587, ... }
    secret 字段若前端回传 "******"，服务端自动跳过，不覆盖真实值。
    """
    if not isinstance(payload, dict):
        raise HTTPException(status_code=400, detail="Payload must be a JSON object")

    # 过滤不在已知 key 列表中的字段
    valid_updates = {k: v for k, v in payload.items() if k in SETTINGS_METADATA}
    if not valid_updates:
        raise HTTPException(status_code=400, detail="No valid settings keys provided")

    _, version = await settings_service.bulk_update(db, valid_updates)
    await db.commit()

    logger.info(f"System settings updated: {list(valid_updates.keys())}, config_version={version}")

    # 异步推送到所有 Region
    background_tasks.add_task(_push_settings_to_all_regions)

    rows = await settings_service.get_all(db)
    return SettingsListResponse(
        settings=[SettingResponse.model_validate(r) for r in rows]
    )


@router.post("/test-notification", response_model=TestNotificationResponse)
async def test_notification(
    req: TestNotificationRequest,
    db: AsyncSession = Depends(get_db),
    _: User = Depends(get_current_superuser),
):
    """测试通知渠道连通性"""
    channel = req.channel
    try:
        if channel == "email":
            result = await _test_email(db)
        elif channel == "feishu":
            result = await _test_feishu(db)
        elif channel == "slack":
            result = await _test_slack(db)
        elif channel == "webhook":
            result = await _test_webhook(db)
        else:
            raise HTTPException(status_code=400, detail=f"Unsupported channel: {channel}")
    except HTTPException:
        raise
    except Exception as e:
        return TestNotificationResponse(channel=channel, success=False, message=str(e))

    return result


# --- Helpers ---

async def _push_settings_to_all_regions():
    """后台任务：将系统设置同步到所有活跃 Region"""
    from app.services.settings_sync_service import settings_sync_service
    async with AsyncSessionLocal() as db:
        try:
            await settings_sync_service.push_settings_to_all_regions(db)
        except Exception as e:
            logger.error(f"Failed to push system settings to regions: {e}")


async def _test_email(db: AsyncSession) -> TestNotificationResponse:
    host = await settings_service.get(db, "SMTP_HOST") or ""
    port = int(await settings_service.get(db, "SMTP_PORT") or 587)
    tls = await settings_service.get(db, "SMTP_TLS")
    user = await settings_service.get(db, "SMTP_USER") or ""
    password = await settings_service.get(db, "SMTP_PASSWORD") or ""
    from_addr = await settings_service.get(db, "SMTP_FROM") or ""
    from_name = await settings_service.get(db, "SMTP_FROM_NAME") or "CloudLand"

    if not host:
        return TestNotificationResponse(channel="email", success=False, message="SMTP_HOST not configured")

    def _send():
        msg = email.mime.text.MIMEText("This is a test notification from CloudLand.", "plain", "utf-8")
        msg["Subject"] = f"[{from_name}] Test Notification"
        msg["From"] = f"{from_name} <{from_addr}>"
        msg["To"] = from_addr

        if port == 465:
            # 隐式 SSL（SMTP_SSL）
            context = ssl.create_default_context()
            with smtplib.SMTP_SSL(host, port, context=context) as smtp:
                if user:
                    smtp.login(user, password)
                smtp.send_message(msg)
        elif tls:
            # 显式 TLS（STARTTLS，常见于端口 587）
            with smtplib.SMTP(host, port) as smtp:
                smtp.ehlo()
                smtp.starttls(context=ssl.create_default_context())
                if user:
                    smtp.login(user, password)
                smtp.send_message(msg)
        else:
            # 明文 SMTP（如内网端口 25）
            with smtplib.SMTP(host, port) as smtp:
                smtp.ehlo()
                if user:
                    smtp.login(user, password)
                smtp.send_message(msg)

    try:
        await asyncio.get_running_loop().run_in_executor(None, _send)
        return TestNotificationResponse(channel="email", success=True, message="Test email sent successfully")
    except Exception as e:
        return TestNotificationResponse(channel="email", success=False, message=str(e))


async def _test_feishu(db: AsyncSession) -> TestNotificationResponse:
    webhook_url = await settings_service.get(db, "FEISHU_WEBHOOK_URL") or ""
    feishu_secret = await settings_service.get(db, "FEISHU_SECRET") or ""
    if not webhook_url:
        return TestNotificationResponse(channel="feishu", success=False, message="FEISHU_WEBHOOK_URL not configured")

    payload: dict = {
        "msg_type": "text",
        "content": {"text": "[CloudLand] Test notification — 飞书渠道连通性测试成功"},
    }

    # 若配置了签名密钥，使用与 notifier.go 相同的 HMAC-SHA256 签名算法进行测试
    if feishu_secret:
        timestamp = str(int(time.time()))
        string_to_sign = f"{timestamp}\n{feishu_secret}"
        mac = hmac_lib.new(string_to_sign.encode("utf-8"), digestmod=hashlib.sha256)
        sign = base64.b64encode(mac.digest()).decode("utf-8")
        payload["timestamp"] = timestamp
        payload["sign"] = sign

    async with httpx.AsyncClient(timeout=10.0) as client:
        resp = await client.post(webhook_url, json=payload)
        if resp.status_code == 200:
            data = resp.json()
            if data.get("code") == 0 or data.get("StatusCode") == 0:
                return TestNotificationResponse(channel="feishu", success=True, message="Feishu webhook OK")
            return TestNotificationResponse(channel="feishu", success=False, message=f"Feishu returned: {data}")
        return TestNotificationResponse(channel="feishu", success=False, message=f"HTTP {resp.status_code}")


async def _test_slack(db: AsyncSession) -> TestNotificationResponse:
    webhook_url = await settings_service.get(db, "SLACK_WEBHOOK_URL") or ""
    if not webhook_url:
        return TestNotificationResponse(channel="slack", success=False, message="SLACK_WEBHOOK_URL not configured")

    payload = {
        "text": "[CloudLand] Test notification — Slack channel connectivity test successful",
    }
    async with httpx.AsyncClient(timeout=10.0) as client:
        resp = await client.post(webhook_url, json=payload)
        if resp.status_code == 200:
            return TestNotificationResponse(channel="slack", success=True, message="Slack webhook OK")
        return TestNotificationResponse(channel="slack", success=False, message=f"HTTP {resp.status_code}: {resp.text[:200]}")


async def _test_webhook(db: AsyncSession) -> TestNotificationResponse:
    url = await settings_service.get(db, "CUSTOM_WEBHOOK_URL") or ""
    method = (await settings_service.get(db, "CUSTOM_WEBHOOK_METHOD") or "POST").upper()
    headers_raw = await settings_service.get(db, "CUSTOM_WEBHOOK_HEADERS") or {}
    # headers 可能以 JSON 字符串形式存储，确保解析为 dict
    if isinstance(headers_raw, str):
        try:
            headers_raw = json.loads(headers_raw)
        except Exception:
            headers_raw = {}
    if not isinstance(headers_raw, dict):
        headers_raw = {}

    if not url:
        return TestNotificationResponse(channel="webhook", success=False, message="CUSTOM_WEBHOOK_URL not configured")

    # 仅允许合法 HTTP 方法，防止注入
    allowed_methods = {"GET", "POST", "PUT", "PATCH"}
    if method not in allowed_methods:
        return TestNotificationResponse(channel="webhook", success=False, message=f"Unsupported HTTP method: {method}")

    payload = {"event": "test", "message": "CloudLand webhook connectivity test"}
    async with httpx.AsyncClient(timeout=10.0) as client:
        resp = await client.request(method, url, json=payload, headers=headers_raw)
        if resp.status_code < 300:
            return TestNotificationResponse(channel="webhook", success=True, message=f"Webhook OK (HTTP {resp.status_code})")
        return TestNotificationResponse(channel="webhook", success=False, message=f"HTTP {resp.status_code}: {resp.text[:200]}")

File: cpgateway-go/src/apis/system_settings.go
package apis

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/gin-gonic/gin"
	log "github.com/sirupsen/logrus"
	"go.opentelemetry.io/otel/trace"
	"gorm.io/gorm"

	"cpgateway-go/src/common"
	"cpgateway-go/src/dbs"
	"cpgateway-go/src/services"
	"cpgateway-go/src/tracing"
)

func settingsListResponse(db *gorm.DB) gin.H {
	rows := services.GetAllSettings(db)
	out := make([]settingOut, 0, len(rows))
	for i := range rows {
		out = append(out, toSettingOut(&rows[i]))
	}
	return gin.H{"settings": out}
}

// GET /system/settings (superuser)
func ListSystemSettings(c *gin.Context) {
	c.JSON(http.StatusOK, settingsListResponse(dbs.DBContext(c.Request.Context())))
}

// PUT /system/settings (superuser) — body is {KEY: value, ...}; unknown keys are ignored and
// masked secrets ("******") are not written back.
func UpdateSystemSettings(c *gin.Context) {
	var payload map[string]interface{}
	if err := c.ShouldBindJSON(&payload); err != nil {
		common.AbortValidation(c, "body", err)
		return
	}
	valid := map[string]interface{}{}
	keys := make([]string, 0, len(payload))
	for k, v := range payload {
		if _, ok := services.FindSettingMeta(k); ok {
			valid[k] = v
			keys = append(keys, k)
		}
	}
	if len(valid) == 0 {
		common.AbortWithDetail(c, http.StatusBadRequest, "No valid settings keys provided")
		return
	}

	db := dbs.DBContext(c.Request.Context())
	version, err := services.BulkUpdateSettings(db, valid)
	if err != nil {
		internalServerError(c, err)
		return
	}
	log.WithContext(c).Infof("System settings updated: %v, config_version=%d", keys, version)
	go services.PushSettingsToAllRegions(context.WithoutCancel(c.Request.Context()))
	c.JSON(http.StatusOK, settingsListResponse(db))
}

// POST /system/settings/test-notification (superuser)
func TestNotification(c *gin.Context) {
	var in struct {
		Channel string `json:"channel" binding:"required"`
	}
	if !bindJSON(c, &in) {
		return
	}
	db := dbs.DBContext(c.Request.Context())
	var success bool
	var message string
	switch in.Channel {
	case "email":
		success, message = testEmailChannel(c.Request.Context(), db)
	case "feishu":
		success, message = testFeishuChannel(c.Request.Context(), db)
	case "slack":
		success, message = testSlackChannel(c.Request.Context(), db)
	case "webhook":
		success, message = testWebhookChannel(c.Request.Context(), db)
	default:
		common.AbortWithDetail(c, http.StatusBadRequest, fmt.Sprintf("Unsupported channel: %s", in.Channel))
		return
	}
	c.JSON(http.StatusOK, gin.H{"channel": in.Channel, "success": success, "message": message})
}

func truncate(s string, n int) string {
	if len(s) > n {
		return s[:n]
	}
	return s
}

func testEmailChannel(ctx context.Context, db *gorm.DB) (bool, string) {
	cfg := services.LoadEmailConfig(db)
	if cfg.SMTP.Host == "" {
		return false, "SMTP_HOST not configured"
	}
	msg := common.BuildPlainMessage(fmt.Sprintf("[%s] Test Notification", cfg.FromName),
		common.FormatAddress(cfg.FromName, cfg.From), cfg.From, "This is a test notification from CloudLand.")
	// SMTP 为外部服务：只建 span
	_, mailSpan := tracing.StartChild(ctx, "notify.email", trace.WithSpanKind(trace.SpanKindClient))
	mailErr := common.SendMail(cfg.SMTP, cfg.From, cfg.From, msg)
	tracing.EndSpan(mailSpan, mailErr)
	if err := mailErr; err != nil {
		return false, err.Error()
	}
	return true, "Test email sent successfully"
}

func postJSON(ctx context.Context, method, url string, payload interface{}, headers map[string]string) (int, []byte, error) {
	// 通知测试目标均为外部服务：只建 span，不注入 trace 头
	ctx, span := tracing.StartChild(ctx, "notify.test", trace.WithSpanKind(trace.SpanKindClient))
	defer span.End()
	body, _ := json.Marshal(payload)
	req, err := http.NewRequestWithContext(ctx, method, url, bytes.NewReader(body))
	if err != nil {
		return 0, nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	for k, v := range headers {
		req.Header.Set(k, v)
	}
	resp, err := (&http.Client{Timeout: 10 * time.Second}).Do(req)
	if err != nil {
		return 0, nil, err
	}
	defer resp.Body.Close()
	respBody, _ := io.ReadAll(resp.Body)
	return resp.StatusCode, respBody, nil
}

func testFeishuChannel(ctx context.Context, db *gorm.DB) (bool, string) {
	webhookURL := services.GetSettingString(db, "FEISHU_WEBHOOK_URL", "")
	secret := services.GetSettingString(db, "FEISHU_SECRET", "")
	if webhookURL == "" {
		return false, "FEISHU_WEBHOOK_URL not configured"
	}
	payload := map[string]interface{}{
		"msg_type": "text",
		"content":  map[string]string{"text": "[CloudLand] Test notification — 飞书渠道连通性测试成功"},
	}
	if secret != "" {
		ts := time.Now().Unix()
		payload["timestamp"] = strconv.FormatInt(ts, 10)
		payload["sign"] = services.FeishuSign(secret, ts)
	}
	status, body, err := postJSON(ctx, http.MethodPost, webhookURL, payload, nil)
	if err != nil {
		return false, err.Error()
	}
	if status != http.StatusOK {
		return false, fmt.Sprintf("HTTP %d", status)
	}
	var data map[string]interface{}
	if err := json.Unmarshal(body, &data); err != nil {
		return false, err.Error()
	}
	for _, key := range []string{"code", "StatusCode"} {
		if v, ok := data[key]; ok && v != nil {
			if f, isNum := services.ToFloat(v); isNum && f == 0 {
				return true, "Feishu webhook OK"
			}
		}
	}
	return false, fmt.Sprintf("Feishu returned: %v", data)
}

func testSlackChannel(ctx context.Context, db *gorm.DB) (bool, string) {
	webhookURL := services.GetSettingString(db, "SLACK_WEBHOOK_URL", "")
	if webhookURL == "" {
		return false, "SLACK_WEBHOOK_URL not configured"
	}
	payload := map[string]string{"text": "[CloudLand] Test notification — Slack channel connectivity test successful"}
	status, body, err := postJSON(ctx, http.MethodPost, webhookURL, payload, nil)
	if err != nil {
		return false, err.Error()
	}
	if status == http.StatusOK {
		return true, "Slack webhook OK"
	}
	return false, fmt.Sprintf("HTTP %d: %s", status, truncate(string(body), 200))
}

func testWebhookChannel(ctx context.Context, db *gorm.DB) (bool, string) {
	webhookURL := services.GetSettingString(db, "CUSTOM_WEBHOOK_URL", "")
	method := strings.ToUpper(services.GetSettingString(db, "CUSTOM_WEBHOOK_METHOD", "POST"))

	headers := map[string]string{}
	rawHeaders := services.GetSetting(db, "CUSTOM_WEBHOOK_HEADERS")
	if s, ok := rawHeaders.(string); ok {
		var parsed interface{}
		if json.Unmarshal([]byte(s), &parsed) == nil {
			rawHeaders = parsed
		}
	}
	if m, ok := rawHeaders.(map[string]interface{}); ok {
		for k, v := range m {
			headers[k] = fmt.Sprint(v)
		}
	}

	if webhookURL == "" {
		return false, "CUSTOM_WEBHOOK_URL not configured"
	}
	switch method {
	case http.MethodGet, http.MethodPost, http.MethodPut, http.MethodPatch:
	default:
		return false, fmt.Sprintf("Unsupported HTTP method: %s", method)
	}

	payload := map[string]string{"event": "test", "message": "CloudLand webhook connectivity test"}
	status, body, err := postJSON(ctx, method, webhookURL, payload, headers)
	if err != nil {
		return false, err.Error()
	}
	if status < 300 {
		return true, fmt.Sprintf("Webhook OK (HTTP %d)", status)
	}
	return false, fmt.Sprintf("HTTP %d: %s", status, truncate(string(body), 200))
}
